File: dobby_spark.py
"""
filename: dobby_spark.py
Description: Dobby spark api to retrieve and post room messages
"""
import ast
import json
import logging
import requests
import time

logger = logging.getLogger('dobby')

BEARER = 'YWE5MDE4ZTItNjYyNi00M2YzLTlkMjEtN2NjZjA4MGIxY2EyYzY0ZDQ3YzQtZmM0'
SPARK_ID = u'Bearer' + ' ' + BEARER
SPARK_URL = "https://api.ciscospark.com/v1/"
        
class Dobby_spark():

    def __init__(self):
        pass
        
    def spark_room_msg(self, msgId):
        """GET an http request from Spark API for room text message
        """
        status = ''
        roomData = {'roomId': msgId}
        try:
            status = requests.get(SPARK_URL+'messages/'+msgId,
                      headers={'Content-type': 'application/json', 'Authorization': SPARK_ID},
                      data=roomData            )
        except:
            logger.warning("  spark_room_msg: fail to retrieve roomData=%s", roomData)
        if hasattr(status, 'status_code') and status.status_code == 200:
            room_json = ast.literal_eval(status.text)
            room_msg = room_json['text']
            person_id = room_json['personId']
            logger.info("  spark_room_msg: successfully get room message text=%s", room_msg)
        else:
            logger.warning("  spark_room_msg: invalid roomId failed get message")
        return person_id, room_msg
    
    def spark_get_person(self, personId):
        """GET an http request from Spark API about person detail.
        """
        status = ''
        personData = {'personId': personId}
        try:
            status = requests.get(SPARK_URL+'people/'+personId, 
                      headers={'Content-type': 'application/json', 'Authorization': SPARK_ID},
                      data=personData)
        except:
            logger.warning("  spark_get_person: fail to post personData=%s", personData)
        if hasattr(status, 'status_code') and status.status_code == 200:
            person_json = ast.literal_eval(status.text)
            name = person_json['displayName']
            logger.info("  spark_get_person: successfully got person name=%s", name)
        else:
            logger.warning("  spark_get_person: invalid personId failed to post personData=%s", personData)
        return name
    
    def spark_msg_request(self, roomData):
        """Post an http request to Spark API for messaging dobby response.
        args: str, roomData - custom text message for Spark room
        """
        status = ''
        try:
            status = requests.post(SPARK_URL+'messages', 
                      headers={'Content-type': 'application/json', 'Authorization': SPARK_ID},
                      data=roomData)
        except:
            logger.warning("  spark_msg_request: fail to post roomData=%s", roomData)
        if hasattr(status, 'status_code') and status.status_code == 200:
            logger.info("  spark_msg_request: successfully post roomData=%s", roomData)
        else:
            logger.warning("  spark_msg_request: invalid roomId failed to post roomData=%s", roomData)
    

File: jslib/dobby_spark.js
/* dobby_spark.js */
'use strict';

const request = require('request');

module.exports = {
  fetchMessage: fetchMessage,
  sendMessage: sendMessage,
  whoAmI: whoAmI,
};

const getReq = request.defaults({
  // uri: 'https://api.ciscospark.com/v1/messages/',
  method: 'GET',
  headers: { 'Content-Type': 'application/json', 'Authorization': 'Bearer MTFkN2UwZDQtN2RjYi00NmQ0LWJkZjQtZDcwNjk1MjBkM2VmOThkZDdmODgtZTgx' },
});

function fetchMessage(token, msgId, cb) {
   request.get({
    uri: 'https://api.ciscospark.com/v1/messages/' + msgId,
    headers: {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer ' + token,
    }}, (err, resp, data) => {
      // console.log("fetched message:", data);
      if (cb) {
        cb(err || data.error && data.error.message, data);
      }
  });
};

const postReq = request.defaults({
  uri: 'https://api.ciscospark.com/v1/messages',
  method: 'POST',
});

function sendMessage(token, roomId, text, cb) {
  const opts = {
    headers: { 'Content-Type': 'application/json', 'Authorization': 'Bearer ' + token },
    json: {
      roomId: roomId,
      text: text,
    },
  };
  postReq(opts, function (err, resp, data) {
    if (cb) {
      cb(err || data.error && data.error.message, data);
    }
  });
};

function whoAmI(token, cb) {
  request.get({
    uri: 'https://api.ciscospark.com/v1/people/me',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer ' + token,
    }}, (err, resp, data) => {
      if (cb) {
        if (err || (data.error && data.error.message) || !data.length) {
          cb(null);
        } else {
          const me = JSON.parse(data);
          cb({
            name: me.displayName,
            id: me.id,
            email: me.emails[0],
          });
        }
      }
    });
}